numba/inline_closurecall.py

```diff
@@ -203,19 +246,435 @@ def _replace_returns(blocks, target, return_label):
+                 ir.Expr.call(empty_func, (size_tuple_var,), (), loc=loc), loc))
+
+    # Add back removed just in case they are used by something else
+    for var in removed:
+        stmts.append(_new_definition(func_ir, var, array_var, loc))
+
+    # Add back terminator
+    stmts.append(terminator)
+    # Modify loop_entry
+    loop_entry.body = stmts
+
+    if not (range_def and range_def[0] == 0):
+        # Insert index_var increment to the end of loop header
+        terminator = loop_header.terminator
+        stmts = loop_header.body[0:-1]
+        next_index_var = scope.make_temp(loc)
+        one = scope.make_temp(loc)
+        # one = 1
+        stmts.append(_new_definition(func_ir, one,
+                     ir.Const(value=1,loc=loc), loc))
+        # next_index_var = index_var + 1
+        stmts.append(_new_definition(func_ir, next_index_var,
+                     ir.Expr.binop(fn='+', lhs=index_var, rhs=one, loc=loc), loc))
+        # index_var = next_index_var
+        stmts.append(_new_definition(func_ir, index_var, next_index_var, loc))
+        stmts.append(terminator)
+        loop_header.body = stmts
+
+    # In append_block, change list_append into array assign
+    for i in range(len(append_block.body)):
+        if append_block.body[i] == append_stmt:
+            debug_print("Replace append with SetItem")
+            append_block.body[i] = ir.SetItem(target=array_var, index=index_var,
+                                              value=append_stmt.value.args[0], loc=append_stmt.loc)
+
+    # replace array call, by changing "a = array(b)" to "a = b"
+    stmt = func_ir.blocks[exit_block].body[array_call_index]
+    # stmt can be either array call or SetItem, we only replace array call
+    if isinstance(stmt, ir.Assign) and isinstance(stmt.value, ir.Expr):
+        stmt.value = array_var
+        func_ir._definitions[stmt.target.name] = [stmt.value]
+
+    return True
+
+
+def _fix_nested_array(func_ir):
+    """Look for assignment like: a[..] = b, where both a and b are numpy arrays, and
+    try to eliminate array b by expanding a with an extra dimension.
+    """
+    """
+    cfg = compute_cfg_from_blocks(func_ir.blocks)
+    all_loops = list(cfg.loops().values())
+    def find_nest_level(label):
+        level = 0
+        for loop in all_loops:
+            if label in loop.body:
+                level += 1
+    """
+
+    def find_array_def(arr):
+        """Find numpy array definition such as arr = numpy.empty(...).  If it is
+        arr = b[...], find array defintion of b recursively.
+        """
+        arr_def = func_ir.get_definition(arr)
+        _make_debug_print("find_array_def")(arr, arr_def)
+        if isinstance(arr_def, ir.Expr):
+            if guard(_find_numpy_call, func_ir, arr_def) == 'empty':
+                return arr_def
+            elif arr_def.op == 'getitem':
+                return find_array_def(arr_def.value)
+        raise GuardException
+
+    def fix_array_assign(stmt):
```

Comment:
The situation arises when there is nested comprehension as an argument to the array call.